# Spam Detector

In [1]:
import pandas as pd
import pipeline_utilities as p_util
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

* I think Logistic Regression will perform better because spam has more of features like spelling mistakes, poor grammar, capitals, etc. So I would think the spam versus non-spam groups or clusters could be separated by a straight line. Also dataset looks somewhat balanced and it is not huge. Spam classification decision-making is not too complex for Logistic Regression.

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']

X = data.copy()
X = X.drop(columns='spam')

In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
#from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
#scaler = StandardScaler()

In [7]:
# Fit the Standard Scaler with the training data
#scaler.fit(X_train)

In [8]:
# Scale the training data
#X_train_scaled = scaler.transform(X_train)
#X_train_scaled

In [9]:
# Transforming the test dataset based on the fit from the training dataset
#X_test_scaled = scaler.transform(X_test)
#X_test_scaled
X_train_scaled, X_test_scaled = p_util.scale_data_with_StandardScaler(X_train, X_test)

Scaled X_train data: [[-0.34818601  0.01397811  2.33184107 ... -0.03236106 -0.07243223
   0.68726115]
 [-0.34818601  1.33393565 -0.56923777 ... -0.0701885  -0.17375566
  -0.30560219]
 [-0.34818601 -0.17020201 -0.56923777 ... -0.10102729 -0.1921781
  -0.27500547]
 ...
 [-0.34818601 -0.17020201 -0.56923777 ... -0.08017228 -0.18296688
  -0.17862583]
 [-0.34818601 -0.17020201 -0.56923777 ... -0.12923147 -0.24744543
  -0.44481723]
 [ 0.50206111  0.03700063  0.53117144 ... -0.056544    0.03810241
   0.07226722]]
Scaled X_train data: [[ 3.76098576e-01 -1.70202007e-01  3.71111921e-01 ... -8.98510004e-02
  -1.96783715e-01  5.69688626e-02]
 [-3.48186008e-01 -1.70202007e-01 -5.69237773e-01 ... -4.60332940e-02
  -1.46122001e-01 -3.25490051e-01]
 [ 1.19485506e+00  4.46748014e-02  2.31059839e-01 ...  3.41415575e-03
   6.36831766e-01  4.45677620e+00]
 ...
 [-3.48186008e-01 -1.70202007e-01 -5.69237773e-01 ... -1.20662059e-01
  -2.15206156e-01 -3.69855285e-01]
 [ 4.39079844e-01  2.13506579e-01  4.31134

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [10]:
# Train a Logistic Regression model and print the model score
# Make and save testing predictions with the saved logistic regression model using the test data
# Review the predictions
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`

# All requirements above have been coded into pipeline_utilities python program file

random_state = 1
p_util.logistic_regression_model_generator(X_train_scaled, X_test_scaled, y_train, y_test, random_state)


Logistic Regression Training Data Score: 0.9269565217391305
Logistic Regression Testing Data Score: 0.9322328410078193
Logistic Regression Predictions: [0 1 1 ... 0 1 0]
Logistic Regression Predictions: 0.9322328410078193


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [11]:
# Train a Random Forest Classifier model and print the model score
# Make and save testing predictions with the saved logistic regression model using the test data
# Review the predictions
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.

# All requirements above have been coded into pipeline_utilities python program file

random_state = 1
n_estimators = 100
p_util.random_forest_model_generator(X_train_scaled, X_test_scaled, y_train, y_test, random_state, n_estimators, X.columns)


Random Forest Training Data Score: 0.9991304347826087
Random Forest Testing Data Score: 0.952215464813206
Random Forest Predictions: [0 1 1 ... 0 1 0]
Random Forest Predictions: 0.952215464813206


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

* Both models had over 90% accuracy score, indicating that both models performed reasonably well, but the Random Forest model performed better than the Logistic Regression model by 2 percentage points. The Random Forest model achieved an accuracy of 95.22% with testing data and predictions while the Logistic Regression model achieved a score of 93.22%.
* My prediction that Logistic Regression model would perform better was incorrect!